Applications/DataExplorer/DataView/DiagramView/DiagramList.cpp

```diff
@@ -288,19 +283,38 @@ int DiagramList::readList(const SensorData* data, std::vector<DiagramList*> &lis
     return nLists;
 }
 
-void DiagramList::setList(std::vector< std::pair<QDateTime, float> > coords)
+void DiagramList::truncateToRange(QDateTime const& start, QDateTime const& end)
 {
-    int numberOfDays;
+    float start_secs = static_cast<float>(_startDate.secsTo(start));
+    if (start_secs < 0)
+        start_secs = 0;
+    float end_secs = static_cast<float>(_startDate.secsTo(end));
+    if (end_secs < start_secs)
+        end_secs = _coords.back().first;
+
+    if (start_secs == 0 && end_secs == _coords.back().first)
+        return;
+
+    _coords.erase(std::remove_if(_coords.begin(), _coords.end(),
+        [&](std::pair<float, float> const& c){return (c.first<start_secs || c.first>end_secs);}),
+        _coords.end());
+    _startDate = start;
+    for (auto& c : _coords)
+        c.first -= start_secs;
+    update();
+}
 
+void DiagramList::setList(std::vector< std::pair<QDateTime, float> > coords)
+{
     this->_startDate = coords[0].first;
     _coords.emplace_back(0.0f, coords[0].second);
 
     std::size_t nCoords = coords.size();
     for (std::size_t i = 1; i < nCoords; i++)
     {
-        numberOfDays = this->_startDate.daysTo(coords[i].first);
-        _coords.emplace_back(static_cast<float>(numberOfDays),
-                             coords[i].second);
+        _coords.emplace_back(
+            static_cast<float>(_startDate.daysTo(coords[i].first)),
+            coords[i].second);
     }
 
```

Comment:
Instead of lines 312--318 you could write:
```std::transform(coords.begin()++, coords.end(), std::back_inserter(_coords),
                   [this](auto const& coord) {
                       return std::make_pair(_startDate.daysTo(coord.first),
                                             coord.second);
                   });
```

Reply:
For the suggested solution one has to check if there are at least 2 elements in the vector.

Reply:
Is this significantly faster? Otherwise I'd rather keep my code because it's easier to read.